### ISSUE TITLE
Interaction with define/private and keyword args is fragile

### ISSUE THREAD

**Issue by pmatos** [racket/racket#2232]
Run:
```
#lang racket

(define myc%
  (class object%
    (super-new)

    (define/public (foo x #:arg (k 0))
      (printf "foo~n")
      (foo* x #:arg k))

    (define/private (foo* y #:arg k)
      (printf "foo*~n")
      (bar y (add1 y) (* y 2)))

    (define/private (bar x y z)
      (printf "~a ~a ~a~n" x y z))))

(define c (new myc%))
(send c foo 2)
```

That works just fine, right? It prints '2 3 4'

Now try changing the foo* definition to using define/private as in:
(define/private (foo* y #:arg k) ...)

Nothing... The call enters foo* though but bar is never called!

@rfindler agreed on the ML this is a bug, which seems to have been introduced between 6.6 and 6.7.

**Comment by pmatos** [comment]
Issue appeared between c5868e9a (good) and ddf6985 (bad). Running a bisect.

**Comment by pmatos** [comment]
First bad commit is cc9889d7 from @mflatt .

**Comment by pmatos** [comment]
I look at the code of the first bad commit and I cannot see how this could possibly be the source of the issue but it seems to be. Commit before that is 01988479 which is good.

**Comment by pmatos** [comment]
Bisect script:
```
#! /bin/bash

cd ~/Projects/racket
git clean -xfd
cd racket/src
(./configure && make -j16) || exit 125
racket/racket3m ~/test-2232.rkt
```

Bisect test:
```
#lang racket

(define myc%
  (class object%
    (super-new)

    (define/public (foo x #:arg (k 0))
      (foo* x #:arg k))

    (define/private (foo* y #:arg k)
      (when (void? (bar y y y))
        (error 'foo* "houston, we have a bug!")))

    (define/private (bar x y z)
      (+ x y z))))

(define c (new myc%))
(send c foo 2)
```

**Comment by mflatt** [comment]
Thanks for tracking down a relevant commit. That was very helpful.

While that commit did not introduce the actual bug, it made the buggy part of the letrec-check analysis relevant. Here's a smaller example:

```
#lang racket/base

(define (wrap x) x)

(letrec ([f (wrap (lambda () (g)))]
         [g (let ([g2 (lambda () (letrec ([x x]) x))])
              g2)])
  (f))
```

The problem is related to the right-hand side of `f` potentially using `g` (as far as the analysis can tell) before the right-hand side of `g` has been inspected. Another key ingredient is the propagation of `g2` to `g`. The analysis didn't track the previously noted use of `g` as implying a use of `g2`, so it never looked at that `lambda` body for `g2`. And then it went one step further, discarding the body of `g2`, because that code didn't seem to be reached.

So, in your example, `bar` is reached, but its body has been replaced by `(void)`.

**Comment by mfelleisen** [comment]
(I guess we never finished a soundness proof and Redex didn’t find this kind of thing.)

**Comment by pmatos** [comment]
@mflatt wow, thanks for the explanation. Certainly quite an intricate issue, but at the same time I am quite surprised that given the amount of code using classes in racket itself, this bug has gone hidden for so long.

@mfelleisen was there an attempt at a soundness proof of racket using redex? Do you have a ref for any initial attemps/paper?

**Comment by mflatt** [comment]
Yes, this is certainly one of those "how did this bug hide for so long?" cases.

**Comment by rfindler** [comment]
Do we still have the model that we used for testing with redex somewhere?

**Comment by mfelleisen** [comment]
@rfindler I was thinking of Claire A's model but I think it doesn't apply here. My bad.

**Comment by rfindler** [comment]
I do feel like we did try something like that, tho! I just can't recall the details anymore.
